File: Windows/Discord4KHelper.Windows/SoundClonerService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Security.Cryptography;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading.Tasks;

namespace Discord4KHelper.Windows;

internal sealed record PluginFile(string Path, string Sha256);
internal sealed record PluginManifest(int SchemaVersion, string HelperVersion, string PluginVersion,
    string VencordCommit, List<PluginFile> Files)
{
    internal static readonly JsonSerializerOptions JsonOptions = new() { PropertyNamingPolicy = JsonNamingPolicy.CamelCase };
    internal const string InvalidPackage = "音效外掛封裝驗證失敗，未變更現有安裝。請重新下載或更新 Helper。";
    internal static PluginManifest Read(string path)
    {
        var manifest = JsonSerializer.Deserialize<PluginManifest>(File.ReadAllText(path), JsonOptions)
            ?? throw new InvalidDataException(InvalidPackage);
        manifest.Validate();
        return manifest;
    }
    internal void Validate()
    {
        if (SchemaVersion != 1 || !Regex.IsMatch(VencordCommit ?? "", "^[a-f0-9]{40}$") ||
            !Regex.IsMatch(HelperVersion ?? "", @"^[0-9]{1,5}\.[0-9]{1,5}\.[0-9]{1,5}$") ||
            !Regex.IsMatch(PluginVersion ?? "", @"^[0-9]{1,5}\.[0-9]{1,5}\.[0-9]{1,5}$") ||
            Files is null || Files.Count is < 1 or > 100 ||
            Files.Select(f => f.Path).Distinct(StringComparer.OrdinalIgnoreCase).Count() != Files.Count ||
            Files.Any(f => !Regex.IsMatch(f.Path ?? "", "^dist/[A-Za-z0-9_-][A-Za-z0-9._-]*$") ||
                !Regex.IsMatch(f.Sha256 ?? "", "^[a-f0-9]{64}$")) ||
            new[] { "dist/patcher.js", "dist/preload.js", "dist/renderer.js", "dist/renderer.css", "dist/package.json" }
                .Except(Files.Select(f => f.Path)).Any()) throw new InvalidDataException(InvalidPackage);
    }
    internal static string Hash(string path) => Convert.ToHexString(SHA256.HashData(File.ReadAllBytes(path))).ToLowerInvariant();
    internal void Verify(string directory)
    {
        Validate();
        var paths = Directory.GetFileSystemEntries(System.IO.Path.Combine(directory, "dist"))
            .Select(p => "dist/" + System.IO.Path.GetFileName(p)).ToHashSet();
        if (!paths.SetEquals(Files.Select(f => f.Path))) throw new InvalidDataException(InvalidPackage);
        foreach (var file in Files)
        {
            var path = System.IO.Path.Combine(directory, file.Path);
            if ((File.GetAttributes(path) & (FileAttributes.ReparsePoint | FileAttributes.Directory)) != 0 || Hash(path) != file.Sha256)
                throw new InvalidDataException(InvalidPackage);
        }
    }
    internal bool SameBuild(PluginManifest other) => SchemaVersion == other.SchemaVersion &&
        HelperVersion == other.HelperVersion && PluginVersion == other.PluginVersion && VencordCommit == other.VencordCommit &&
        Files.OrderBy(f => f.Path).SequenceEqual(other.Files.OrderBy(f => f.Path));
}

internal static class SoundClonerService
{
    internal const string ArchiveName = "Vencord-SoundCloner.zip";
    internal const string ManifestName = "Vencord-SoundCloner-manifest.json";
    internal const string MarkerName = ".soundcloner-manifest.json";
    internal static string Backup(string root) => Path.Combine(root, "discord-4k-helper-backup", "dist");
    internal static async Task<PluginManifest> GetManifestAsync(GitHubRelease release)
    {
        var asset = release.Assets.FirstOrDefault(a => a.Name == ManifestName)
            ?? throw new InvalidDataException("這個版本沒有音效外掛封裝，請更新 Helper。");
        var file = Path.Combine(Path.GetTempPath(), $"soundcloner-manifest-{Guid.NewGuid():N}.json");
        try
        {
            await Web.DownloadAsync(asset.DownloadUrl, file);
            var manifest = PluginManifest.Read(file);
            if (AppVersion.Parse(manifest.HelperVersion).CompareTo(release.Version) != 0)
                throw new InvalidDataException(PluginManifest.InvalidPackage);
            return manifest;
        }
        finally { File.Delete(file); }
    }

    internal static async Task<string> PrepareAsync(GitHubRelease release, string helperVersion)
    {
        var expected = await GetManifestAsync(release);
        if (AppVersion.Parse(helperVersion) < AppVersion.Parse(expected.HelperVersion))
            throw new InvalidOperationException("請先更新 Helper，再安裝音效外掛。");
        var asset = release.Assets.FirstOrDefault(a => a.Name == ArchiveName)
            ?? throw new InvalidDataException(PluginManifest.InvalidPackage);
        var stage = Path.Combine(Path.GetTempPath(), $"soundcloner-{Guid.NewGuid():N}");
        Directory.CreateDirectory(stage);
        try
        {
            var archive = Path.Combine(stage, "download.zip");
            await Web.DownloadAsync(asset.DownloadUrl, archive);
            ExtractVerified(archive, stage);
            File.Delete(archive);
            var package = Path.Combine(stage, "Vencord-SoundCloner");
            var manifest = PluginManifest.Read(Path.Combine(package, "manifest.json"));
            if (!manifest.SameBuild(expected)) throw new InvalidDataException(PluginManifest.InvalidPackage);
            manifest.Verify(package);
            return stage;
        }
        catch { Directory.Delete(stage, true); throw; }
    }

    internal static void ExtractVerified(string archive, string stage)
    {
        if (new FileInfo(archive).Length > 128 * 1024 * 1024) throw new InvalidDataException(PluginManifest.InvalidPackage);
        using var zip = ZipFile.OpenRead(archive);
        if (zip.Entries.Count is < 1 or > 110) throw new InvalidDataException(PluginManifest.InvalidPackage);
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        long total = 0;
        foreach (var entry in zip.Entries)
        {
            var name = entry.FullName;
            var kind = (entry.ExternalAttributes >> 16) & 0xf000;
            var allowed = new[] { "Vencord-SoundCloner/", "Vencord-SoundCloner/dist/", "Vencord-SoundCloner/manifest.json", "Vencord-SoundCloner/Vencord-LICENSE.txt" };
            total += entry.Length;
            if ((!allowed.Contains(name) && !Regex.IsMatch(name, "^Vencord-SoundCloner/dist/[A-Za-z0-9_-][A-Za-z0-9._-]*$")) ||
                Regex.IsMatch(Path.GetFileName(name), @"^(CON|PRN|AUX|NUL|COM[0-9]|LPT[0-9])(\.|$)", RegexOptions.IgnoreCase) || name.EndsWith('.') ||
                !seen.Add(name) || kind is not (0 or 0x4000 or 0x8000) ||
                (kind == 0x4000 && !name.EndsWith('/')) || (entry.ExternalAttributes & 0x400) != 0 || total > 128 * 1024 * 1024)
                throw new InvalidDataException(PluginManifest.InvalidPackage);
        }
        // Only exact allowlisted paths are used for extraction; never trust archive paths verbatim.
        foreach (var entry in zip.Entries)
        {
            var destination = Path.Combine(stage, entry.FullName);
            if (entry.FullName.EndsWith('/')) Directory.CreateDirectory(destination);
            else { Directory.CreateDirectory(Path.GetDirectoryName(destination)!); entry.ExtractToFile(destination); }
        }
    }

    internal static PluginManifest? Installed(string root)
    {
        try
        {
            var manifest = PluginManifest.Read(Path.Combine(root, "dist", MarkerName));
            return manifest.Files.All(f => PluginManifest.Hash(Path.Combine(root, f.Path)) == f.Sha256) ? manifest : null;
        }
        catch { return null; }
    }

    internal static bool NeedsCustomWarning(string root)
    {
        if (Installed(root) is not null || !File.Exists(Path.Combine(root, "dist", "patcher.js"))) return false;
        try
        {
            using var map = JsonDocument.Parse(File.ReadAllText(Path.Combine(root, "dist", "renderer.js.map")));
            return map.RootElement.GetProperty("sources").EnumerateArray()
                .Any(s => s.GetString()!.Replace('\\', '/').Contains("/userplugins/"));
        }
        catch { return true; }
    }

    internal static void Install(string stage, string root, string settings, bool enableFeatures = true)
    {
        var package = Path.Combine(stage, "Vencord-SoundCloner");
        var marker = File.ReadAllText(Path.Combine(package, "manifest.json"));
        PluginManifest.Read(Path.Combine(package, "manifest.json")).Verify(package);
        var original = File.Exists(settings) ? File.ReadAllText(settings) : "{}";
        var updated = enableFeatures ? VencordSettings.Edit(original, enabled: true, soundCloner: true) : original;
        var backup = Backup(root);
        if (!Directory.Exists(backup))
        {
            Directory.CreateDirectory(Path.GetDirectoryName(backup)!);
            var temp = Path.Combine(root, $"backup-{Guid.NewGuid():N}");
            try { CopyDirectory(Path.Combine(root, "dist"), temp); Directory.Move(temp, backup); }
            finally { if (Directory.Exists(temp)) Directory.Delete(temp, true); }
        }
        Swap(Path.Combine(package, "dist"), root, settings, updated, marker);
    }

    internal static void Restore(string root, string settings)
    {
        var backup = Backup(root);
        if (!File.Exists(Path.Combine(backup, "patcher.js"))) throw new InvalidOperationException("找不到安裝前的 Vencord 備份，無法自動還原。");
        Swap(backup, root, settings, VencordSettings.Edit(File.Exists(settings) ? File.ReadAllText(settings) : "{}", removeSoundCloner: true), null);
    }

    internal static void Swap(string source, string root, string settings, string updated, string? marker)
    {
        var candidate = Path.Combine(root, $"dist-new-{Guid.NewGuid():N}");
        var previous = Path.Combine(root, $"dist-old-{Guid.NewGuid():N}");
        var target = Path.Combine(root, "dist");
        try
        {
            CopyDirectory(source, candidate);
            if (marker is not null) File.WriteAllText(Path.Combine(candidate, MarkerName), marker);
            Directory.CreateDirectory(Path.GetDirectoryName(settings)!);
            Directory.Move(target, previous);
            try
            {
                Directory.Move(candidate, target);
                VencordSettings.WriteAtomic(settings, updated);
            }
            catch
            {
                if (Directory.Exists(target)) Directory.Delete(target, true);
                Directory.Move(previous, target);
                throw;
            }
            try { Directory.Delete(previous, true); } catch { /* A retained rollback copy is safe. */ }
        }
        finally { if (Directory.Exists(candidate)) Directory.Delete(candidate, true); }
    }

    private static void CopyDirectory(string source, string target)
    {
        if ((File.GetAttributes(source) & FileAttributes.ReparsePoint) != 0)
            throw new InvalidDataException("Vencord 資料夾含符號連結，請先手動備份再安裝。");
        Directory.CreateDirectory(target);
        foreach (var path in Directory.EnumerateFileSystemEntries(source))
        {
            if ((File.GetAttributes(path) & FileAttributes.ReparsePoint) != 0)
                throw new InvalidDataException("Vencord 資料夾含符號連結，請先手動備份再安裝。");
            var destination = Path.Combine(target, Path.GetFileName(path));
            if (Directory.Exists(path)) CopyDirectory(path, destination);
            else File.Copy(path, destination);
        }
    }
}

File: Sources/Discord4KHelper/SoundClonerService.swift
import CryptoKit
import Foundation

struct PluginManifest: Codable, Equatable {
    struct File: Codable, Equatable {
        let path: String
        let sha256: String
    }
    let schemaVersion: Int
    let helperVersion: String
    let pluginVersion: String
    let vencordCommit: String
    let files: [File]

    func validate() throws {
        guard schemaVersion == 1,
              helperVersion.range(of: "^[0-9]{1,5}\\.[0-9]{1,5}\\.[0-9]{1,5}$", options: .regularExpression) != nil,
              pluginVersion.range(of: "^[0-9]{1,5}\\.[0-9]{1,5}\\.[0-9]{1,5}$", options: .regularExpression) != nil,
              vencordCommit.range(of: "^[a-f0-9]{40}$", options: .regularExpression) != nil,
              !files.isEmpty, files.count <= 100,
              Set(files.map { $0.path.lowercased() }).count == files.count,
              Set(["dist/patcher.js", "dist/preload.js", "dist/renderer.js", "dist/renderer.css", "dist/package.json"])
                .isSubset(of: Set(files.map(\.path))),
              files.allSatisfy({ file in
                  file.path.range(of: "^dist/[A-Za-z0-9_-][A-Za-z0-9._-]*$", options: .regularExpression) != nil
                    && file.sha256.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil
              })
        else { throw HelperError.invalidPluginPackage }
    }

    static func hash(_ data: Data) -> String {
        SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
    }

    func verify(directory: URL) throws {
        try validate()
        let fm = FileManager.default
        let actual = try fm.contentsOfDirectory(atPath: directory.appendingPathComponent("dist").path)
        guard Set(actual.map { "dist/" + $0 }) == Set(files.map(\.path)) else {
            throw HelperError.invalidPluginPackage
        }
        for file in files {
            let url = directory.appendingPathComponent(file.path)
            let attrs = try fm.attributesOfItem(atPath: url.path)
            guard attrs[.type] as? FileAttributeType == .typeRegular,
                  Self.hash(try Data(contentsOf: url)) == file.sha256 else {
                throw HelperError.invalidPluginPackage
            }
        }
    }
}

// Validate central AND local names before invoking the system ZIP extractor.
// This format deliberately only supports a small flat, non-ZIP64 package.
enum PluginZip {
    static func validate(_ data: Data) throws {
        let bytes = [UInt8](data)
        func number(_ offset: Int, _ length: Int) throws -> Int {
            guard offset >= 0, offset + length <= bytes.count else { throw HelperError.invalidPluginPackage }
            return (0..<length).reduce(0) { $0 | Int(bytes[offset + $1]) << ($1 * 8) }
        }
        guard bytes.count >= 22, bytes.count <= 128 * 1024 * 1024 else { throw HelperError.invalidPluginPackage }
        var end: Int?
        for offset in stride(from: bytes.count - 22, through: max(0, bytes.count - 65557), by: -1) {
            if try number(offset, 4) == 0x06054b50,
               try offset + 22 + number(offset + 20, 2) == bytes.count { end = offset; break }
        }
        guard let end, try number(end + 4, 4) == 0 else { throw HelperError.invalidPluginPackage }
        let count = try number(end + 10, 2)
        guard count > 0, count <= 110, try number(end + 8, 2) == count else { throw HelperError.invalidPluginPackage }
        var cursor = try number(end + 16, 4)
        guard try cursor + number(end + 12, 4) == end else { throw HelperError.invalidPluginPackage }
        var seen = Set<String>()
        var total = 0
        for _ in 0..<count {
            guard try number(cursor, 4) == 0x02014b50,
                  try number(cursor + 8, 2) & 1 == 0,
                  try [0, 8].contains(number(cursor + 10, 2)),
                  try number(cursor + 34, 2) == 0 else { throw HelperError.invalidPluginPackage }
            let nameLength = try number(cursor + 28, 2)
            let nameStart = cursor + 46
            guard nameLength > 0, nameStart + nameLength <= end else { throw HelperError.invalidPluginPackage }
            let name = String(decoding: bytes[nameStart..<nameStart + nameLength], as: UTF8.self)
            let allowed = ["Vencord-SoundCloner/", "Vencord-SoundCloner/dist/", "Vencord-SoundCloner/manifest.json", "Vencord-SoundCloner/Vencord-LICENSE.txt"]
            guard allowed.contains(name) || name.range(of: "^Vencord-SoundCloner/dist/[A-Za-z0-9_-][A-Za-z0-9._-]*$", options: .regularExpression) != nil,
                  seen.insert(name.lowercased()).inserted else { throw HelperError.invalidPluginPackage }
            let kind = try (number(cursor + 38, 4) >> 16) & 0xf000
            guard [0, 0x4000, 0x8000].contains(kind),
                  kind != 0x4000 || name.hasSuffix("/") else { throw HelperError.invalidPluginPackage }
            total += try number(cursor + 24, 4)
            guard total <= 128 * 1024 * 1024 else { throw HelperError.invalidPluginPackage }
            let local = try number(cursor + 42, 4)
            guard try number(local, 4) == 0x04034b50,
                  try number(local + 26, 2) == nameLength,
                  try number(local + 6, 2) == number(cursor + 8, 2),
                  try number(local + 8, 2) == number(cursor + 10, 2),
                  local + 30 + nameLength <= cursor else { throw HelperError.invalidPluginPackage }
            let localName = String(decoding: bytes[local + 30..<local + 30 + nameLength], as: UTF8.self)
            guard localName == name else { throw HelperError.invalidPluginPackage }
            cursor += try 46 + nameLength + number(cursor + 30, 2) + number(cursor + 32, 2)
        }
        guard cursor == end else { throw HelperError.invalidPluginPackage }
    }
}

enum SoundClonerService {
    static let archiveName = "Vencord-SoundCloner.zip"
    static let manifestName = "Vencord-SoundCloner-manifest.json"
    static let markerName = ".soundcloner-manifest.json"
    static let fm = FileManager.default

    static func manifest(for release: GitHubRelease) async throws -> PluginManifest {
        guard let asset = release.asset(named: manifestName) else { throw HelperError.missingUpdateAsset }
        let file = try await NetworkService.download(asset.downloadURL)
        defer { try? fm.removeItem(at: file) }
        let manifest = try JSONDecoder().decode(PluginManifest.self, from: Data(contentsOf: file))
        try manifest.validate()
        guard AppVersion(manifest.helperVersion) == release.version else { throw HelperError.invalidPluginPackage }
        return manifest
    }

    static func prepare(_ release: GitHubRelease, helperVersion: String) async throws -> URL {
        let expected = try await manifest(for: release)
        guard let current = AppVersion(helperVersion), let required = AppVersion(expected.helperVersion),
              current >= required, let asset = release.asset(named: archiveName) else { throw HelperError.invalidPluginPackage }
        let archive = try await NetworkService.download(asset.downloadURL)
        defer { try? fm.removeItem(at: archive) }
        try PluginZip.validate(Data(contentsOf: archive))
        let stage = fm.temporaryDirectory.appendingPathComponent("soundcloner-\(UUID().uuidString)")
        try fm.createDirectory(at: stage, withIntermediateDirectories: false)
        do {
            _ = try await ProcessRunner.run(URL(fileURLWithPath: "/usr/bin/ditto"), arguments: ["-x", "-k", archive.path, stage.path])
            let package = stage.appendingPathComponent("Vencord-SoundCloner")
            let manifest = try JSONDecoder().decode(PluginManifest.self, from: Data(contentsOf: package.appendingPathComponent("manifest.json")))
            guard manifest == expected else { throw HelperError.invalidPluginPackage }
            try manifest.verify(directory: package)
            return stage
        } catch { try? fm.removeItem(at: stage); throw error }
    }

    static func installed(in root: URL) -> PluginManifest? {
        guard let data = try? Data(contentsOf: root.appendingPathComponent("dist/\(markerName)")),
              let manifest = try? JSONDecoder().decode(PluginManifest.self, from: data),
              (try? manifest.validate()) != nil,
              manifest.files.allSatisfy({ file in
                  guard let bytes = try? Data(contentsOf: root.appendingPathComponent(file.path)) else { return false }
                  return PluginManifest.hash(bytes) == file.sha256
              }) else { return nil }
        return manifest
    }

    static func needsCustomWarning(in root: URL) -> Bool {
        guard installed(in: root) == nil, fm.fileExists(atPath: root.appendingPathComponent("dist/patcher.js").path) else { return false }
        guard let data = try? Data(contentsOf: root.appendingPathComponent("dist/renderer.js.map")),
              let map = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let sources = map["sources"] as? [String] else { return true }
        return sources.contains { $0.replacingOccurrences(of: "\\", with: "/").contains("/userplugins/") }
    }

    static func backup(in root: URL) -> URL { root.appendingPathComponent("discord-4k-helper-backup/dist") }

    static func install(stage: URL, root: URL, settings: URL, enableFeatures: Bool = true) throws {
        let package = stage.appendingPathComponent("Vencord-SoundCloner")
        let data = try Data(contentsOf: package.appendingPathComponent("manifest.json"))
        let manifest = try JSONDecoder().decode(PluginManifest.self, from: data)
        try manifest.verify(directory: package)
        let original = fm.fileExists(atPath: settings.path) ? try Data(contentsOf: settings) : Data("{}".utf8)
        let updated = enableFeatures ? try SettingsEditor.updating(original, enabled: true, soundCloner: true) : original
        let backupURL = backup(in: root)
        if !fm.fileExists(atPath: backupURL.path) {
            try fm.createDirectory(at: backupURL.deletingLastPathComponent(), withIntermediateDirectories: true)
            let temporaryBackup = root.appendingPathComponent("backup-\(UUID().uuidString)")
            defer { try? fm.removeItem(at: temporaryBackup) }
            try fm.copyItem(at: root.appendingPathComponent("dist"), to: temporaryBackup)
            try fm.moveItem(at: temporaryBackup, to: backupURL)
        }
        try swap(source: package.appendingPathComponent("dist"), root: root, settings: settings, updated: updated, marker: data)
    }

    static func restore(root: URL, settings: URL) throws {
        let source = backup(in: root)
        guard fm.fileExists(atPath: source.appendingPathComponent("patcher.js").path) else { throw HelperError.missingBackup }
        let original = fm.fileExists(atPath: settings.path) ? try Data(contentsOf: settings) : Data("{}".utf8)
        try swap(source: source, root: root, settings: settings,
                 updated: SettingsEditor.updating(original, removeSoundCloner: true), marker: nil)
    }

    // A failed settings write also rolls back dist. The original backup is never overwritten.
    static func swap(source: URL, root: URL, settings: URL, updated: Data, marker: Data?) throws {
        let candidate = root.appendingPathComponent("dist-new-\(UUID().uuidString)")
        let previous = root.appendingPathComponent("dist-old-\(UUID().uuidString)")
        let target = root.appendingPathComponent("dist")
        defer { try? fm.removeItem(at: candidate) }
        try fm.copyItem(at: source, to: candidate)
        if let marker { try marker.write(to: candidate.appendingPathComponent(markerName), options: .atomic) }
        try fm.createDirectory(at: settings.deletingLastPathComponent(), withIntermediateDirectories: true)
        try fm.moveItem(at: target, to: previous)
        do {
            try fm.moveItem(at: candidate, to: target)
            try updated.write(to: settings, options: .atomic)
        } catch {
            if fm.fileExists(atPath: target.path) { try fm.removeItem(at: target) }
            try fm.moveItem(at: previous, to: target)
            throw error
        }
        try? fm.removeItem(at: previous)
    }
}
